### PR TITLE
K8SPXC-767: Don't allow backups to run simultaneously

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPXC-767](https://jira.percona.com/browse/K8SPXC-767) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | failed |
| scaling | passed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| self-healing-advanced-chaos | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing-chaos | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/a4dfb93f67cdc31256e81d79f87c7fa84e3c2c63
image: `perconalab/percona-xtradb-cluster-operator:PR-916-a4dfb93f`

File: pkg/controller/pxcbackup/controller.go [hunk 1]
```diff
@@ -495,3 +504,26 @@ func setControllerReference(cr *api.PerconaXtraDBClusterBackup, obj metav1.Objec
...
+		}
+		return false, errors.Wrap(err, "failed to get backup list")
+	}
+
+	for _, bcp := range bcpList.Items {
+		if bcp.Spec.PXCCluster != cluster.Name {
+			continue
+		}
+
+		if bcp.Name != cr.Name && (bcp.Status.State == api.BackupRunning || bcp.Status.State == api.BackupStarting) {
```
Comment (delgod) [inline pkg/controller/pxcbackup/controller.go:523]: it is better to check only final/finished states
reason: we can add more statuses in the future
```suggestion
		if bcp.Name != cr.Name && bcp.Status.State != api.BackupFailed && bcp.Status.State != api.BackupSucceeded) {
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by qJkee — DISMISSED** [review]

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| storage | passed |
| scheduled-backup | failed |
| recreate | passed |
| big-data | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | failed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| limits | passed |
| self-healing-advanced | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| smart-update | failed |
| affinity | passed |
| self-healing-advanced-chaos | passed |
| one-pod | passed |
| pitr | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing-chaos | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/a720588ba9ad248f4f3f81590b70ac91ca79f798
image: `perconalab/percona-xtradb-cluster-operator:PR-916-a720588b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| haproxy | passed |
| init-deploy | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| scheduled-backup | failed |
| scaling-proxysql | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/a720588ba9ad248f4f3f81590b70ac91ca79f798
image: `perconalab/percona-xtradb-cluster-operator:PR-916-a720588b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| scheduled-backup | failed |
| recreate | passed |
| storage | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced-chaos | failed |
| pitr | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| operator-self-healing-chaos | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/8634e3df4fc48f2f0259b410eafc601fb68c166f
image: `perconalab/percona-xtradb-cluster-operator:PR-916-8634e3df`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| big-data | passed |
| init-deploy | passed |
| limits | passed |
| storage | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| affinity | passed |
| self-healing-chaos | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| recreate | passed |
| scaling | passed |
| self-healing-advanced-chaos | passed |
| restore-to-encrypted-cluster | passed |
| proxysql-sidecar-res-limits | passed |
| scaling-proxysql | passed |
| demand-backup | passed |
| users | passed |
| security-context | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-haproxy | passed |
| scheduled-backup | failed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/8634e3df4fc48f2f0259b410eafc601fb68c166f
image: `perconalab/percona-xtradb-cluster-operator:PR-916-8634e3df`

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| scheduled-backup | failed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| haproxy | passed |
| scaling | passed |
| storage | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| security-context | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| self-healing-advanced-chaos | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| operator-self-healing-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/916/commits/8634e3df4fc48f2f0259b410eafc601fb68c166f
image: `perconalab/percona-xtradb-cluster-operator:PR-916-8634e3df`

**Comment by hors** [comment]
@egegunes  I do not understand why we should block it. E.g we have 5 nodes why we can't perform 2 backups at the same time... We have the feature that we don't run the backups on the primary e.g. if 3 nodes are available(pxc-0,pxc-1 and pxc-2)   only pxc-1 and pxc-2 are ok for backups. And after these changes we will have a big issue with our scheduled-backup test because we run a lot of backups simultaneously in this test. @delgod  what do you think about it?

**Review by hors — CHANGES_REQUESTED** [review]
please see comment

**Comment by egegunes** [comment]
@hors Honestly, I agree with you. The changes are for: https://jira.percona.com/browse/K8SPXC-767?focusedCommentId=283540&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-283540

**Review by delgod — CHANGES_REQUESTED** [review]
slava is right, allowed amount of parallel backups should be qual to spec.pxc.size

**Comment by egegunes** [comment]
@delgod @hors 

If you have a 3 node PXC cluster and create like 9 backup CRs, all backups start at the same time. But the backup image fails if a backup already in progress and you got something like this:

```
NAME                                               READY   STATUS      RESTARTS   AGE
milleporine-15129-haproxy-0                        2/2     Running     2          53m
milleporine-15129-haproxy-1                        2/2     Running     0          52m
milleporine-15129-haproxy-2                        2/2     Running     2          52m
milleporine-15129-pxc-0                            3/3     Running     0          53m
milleporine-15129-pxc-1                            3/3     Running     0          52m
milleporine-15129-pxc-2                            3/3     Running     0          52m
percona-xtradb-cluster-operator-744d87b699-r5khb   1/1     Running     0          61m
xb-backup10-9nn6t                                  0/1     Completed   0          46m
xb-backup10-wpsbf                                  0/1     Error       0          47m
xb-backup11-76m9t                                  0/1     Completed   0          45m
xb-backup11-kxm4k                                  0/1     Error       0          47m
xb-backup11-vqlg7                                  0/1     Error       0          46m
xb-backup12-xw8lp                                  0/1     Completed   0          46m
xb-backup13-8tmjj                                  0/1     Error       0          46m
xb-backup13-glrpk                                  0/1     Error       0          46m
xb-backup13-ktphp                                  0/1     Error       0          46m
xb-backup13-tq8mv                                  0/1     Completed   0          44m
xb-backup13-wlrt5                                  0/1     Error       0          45m
xb-backup14-kdkn9                                  0/1     Completed   0          46m
xb-backup15-b5k9v                                  0/1     Error       0          46m
xb-backup15-h5dlz                                  0/1     Completed   0          46m
xb-backup8-c6fgx                                   0/1     Completed   0          47m
xb-backup9-4sqzx                                   0/1     Error       0          46m
xb-backup9-5f2rl                                   0/1     Completed   0          44m
xb-backup9-nrdnx                                   0/1     Error       0          47m
xb-backup9-wgs6r                                   0/1     Error       0          46m
xb-backup9-wtfsn                                   0/1     Error       0          46m
```

So, we practically have the check to limit running backups to PXC size. The errored backups eventually succeeds because of the job retries. By limiting the running backups on operator level we can get rid of that pods with Error state.

As discussed with @spron-in , I'm closing this for now.

**Pull request closed without merging**